File: iosApp/iosApp/Screens/ScheduleScreen.swift
import SwiftUI

/// 일정.
///
/// 달력은 원래 홈에 있었는데 **일정으로 옮겼다** (2026-09-08 대표 지시).
/// 홈은 다른 화면이 다 찬 뒤에 마지막으로 짠다.
///
/// ⚠️ **헤더만 고정이고 달력부터 아래는 전부 스크롤한다.**
/// `safeAreaInset` 이나 `Group` 으로 감싸면 하단 유리 바가 **끝까지 내렸을 때
/// 제멋대로 펴진다** (2026-09-07 에 홈에서 겪었다).
///
/// 짜임새는 팀버핏 코치 앱을 참고했다 (2026-09-08 대표 지시) —
/// 큰 날짜 제목 → 주 달력 → 그날 수업 카드. 값은 TeamFIS 것(PT 1:1)으로 갈았다.
struct ScheduleScreen: View {
    var onClass: (ScheduleClass) -> Void = { _ in }
    var onNotification: () -> Void = {}
    var onMy: () -> Void = {}

    @State private var selected = Date()
    @State private var month = Date()
    @State private var expanded = false

    var body: some View {
        VStack(spacing: 0) {
            AppHeader(onNotification: onNotification, onMy: onMy)

            ScrollView {
                VStack(alignment: .leading, spacing: 0) {
                    ScheduleCalendar(
                        selected: $selected,
                        month: $month,
                        expanded: expanded,
                        hasClass: Self.hasClass
                    )
                    .padding(.top, TeamFisSpacing.sm)

                    CalendarBar(expanded: expanded) {
                        withAnimation(TeamFisMotion.slow) { expanded.toggle() }
                    }
                    .padding(.top, TeamFisSpacing.xs)

                    let classes = Self.classes(on: selected)
                    if classes.isEmpty {
                        Text("일정이 없어요")
                            .font(TeamFisFont.bodySm)
                            .foregroundStyle(TeamFisColor.textMuted)
                            .frame(maxWidth: .infinity)
                            .padding(.top, TeamFisSpacing.xxxl)
                    } else {
                        VStack(spacing: TeamFisSpacing.md) {
                            ForEach(classes) { item in
                                ScheduleClassCard(item: item, onSelect: { onClass(item) })
                            }
                        }
                        .padding(.top, TeamFisSpacing.lg)
                        .padding(.horizontal, TeamFisSpacing.screenHorizontal)
                    }
                }
                // 접힌 유리 바가 마지막 줄을 덮지 않게 (값의 근거는 토큰 주석에)
                .padding(.bottom, TeamFisSize.bottomBarClearance)
            }
        }
        .onChange(of: selected) { _, new in month = new }
    }

    /// 데이터가 붙기 전까지 쓰는 **자리 표시자**다. 서버가 일정을 주면 통째로 걷어낸다.
    /// 오늘·내일·사흘 뒤에만 수업이 있는 것으로 둔다 — 점이 찍히는 날과 목록이 어긋나면 안 된다.
    static func hasClass(_ date: Date) -> Bool {
        let calendar = Calendar.current
        let today = Date()
        return [0, 1, 3].contains {
            guard let day = calendar.date(byAdding: .day, value: $0, to: today) else { return false }
            return calendar.isDate(day, inSameDayAs: date)
        }
    }

    static func classes(on date: Date) -> [ScheduleClass] {
        guard hasClass(date) else { return [] }
        return [
            ScheduleClass(member: "000", at: at(date, 10, 0), minutes: 60,
                          product: "얼리버드 20회", progress: "12/20회차", status: .done),
            ScheduleClass(member: "000", at: at(date, 14, 0), minutes: 60,
                          product: "PT 30회", progress: "12/30회차", status: .scheduled),
            ScheduleClass(member: "000", at: at(date, 16, 0), minutes: 60,
                          product: "PT 20회", progress: "3/20회차", status: .scheduled),
            ScheduleClass(member: "000", at: at(date, 18, 30), minutes: 60,
                          product: "얼리버드 10회", progress: "8/10회차", status: .scheduled),
        ]
    }

    private static func at(_ date: Date, _ hour: Int, _ minute: Int) -> Date {
        TeamFisCalendar.calendar.date(
            bySettingHour: hour, minute: minute, second: 0, of: date
        ) ?? date
    }
}

/// 캘린더 아래 한 줄 — `펼쳐보기`.
///
/// 펼치기는 **화살표가 뒤집히며** 캘린더가 그 달로 늘어난다.
private struct CalendarBar: View {
    let expanded: Bool
    let onToggle: () -> Void

    var body: some View {
        HStack(spacing: 0) {
            Button(action: onToggle) {
                HStack(spacing: 2) {
                    Text(expanded ? "접기" : "펼쳐보기")
                        .font(TeamFisFont.bodySm)
                    Image("ic_chevron_down")
                        .renderingMode(.template)
                        .resizable()
                        .frame(width: 18, height: 18)
                        .rotationEffect(.degrees(expanded ? 180 : 0))
                        .animation(TeamFisMotion.base, value: expanded)
                }
                .foregroundStyle(TeamFisColor.textSecondary)
                .padding(.horizontal, TeamFisSpacing.sm)
                .frame(height: TeamFisSize.chip)
                .contentShape(Rectangle())
            }
            .buttonStyle(.plain)

            Spacer(minLength: 0)
        }
        .padding(.horizontal, TeamFisSpacing.screenHorizontal - TeamFisSpacing.sm)
    }
}

File: androidApp/src/main/kotlin/com/teamfis/app/ui/screens/ScheduleScreen.kt
package com.teamfis.app.ui.screens

import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.clickable
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Icon
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.graphicsLayer
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.teamfis.app.R
import com.teamfis.app.ui.components.ScheduleCalendar
import com.teamfis.app.ui.components.ScheduleClass
import com.teamfis.app.ui.components.ScheduleClassCard
import com.teamfis.app.ui.components.SessionStatus
import com.teamfis.app.ui.shell.AppHeader
import com.teamfis.app.ui.theme.TeamFisColor
import com.teamfis.app.ui.theme.TeamFisMotion
import com.teamfis.app.ui.theme.TeamFisRadius
import com.teamfis.app.ui.theme.TeamFisSize
import com.teamfis.app.ui.theme.TeamFisSpacing
import com.teamfis.app.ui.theme.TeamFisType

/**
 * 일정.
 *
 * 달력은 원래 홈에 있었는데 **일정으로 옮겼다** (2026-09-08 대표 지시).
 * 홈은 다른 화면이 다 찬 뒤에 마지막으로 짠다.
 *
 * **헤더만 고정이고 날짜 제목부터 아래는 전부 스크롤한다.**
 *
 * 짜임새는 팀버핏 코치 앱을 참고했다 (2026-09-08 대표 지시) —
 * 큰 날짜 제목 → 주 달력 → 그날 수업 카드. 값은 TeamFIS 것(PT 1:1)으로 갈았다.
 */
@Composable
fun ScheduleScreen(
    onClass: (ScheduleClass) -> Unit = {},
    onNotification: () -> Unit = {},
    onMy: () -> Unit = {},
) {
    var selected by remember { mutableStateOf(java.time.LocalDate.now()) }
    var month by remember { mutableStateOf(java.time.LocalDate.now()) }
    var expanded by rememberSaveable { mutableStateOf(false) }

    Column(Modifier.fillMaxSize()) {
        AppHeader(onNotification = onNotification, onMy = onMy)

        Column(
            Modifier
                .fillMaxSize()
                .verticalScroll(rememberScrollState())
                .padding(bottom = TeamFisSpacing.xxxl),
        ) {
            ScheduleCalendar(
                selected = selected,
                month = month,
                expanded = expanded,
                onSelect = {
                    selected = it
                    month = it
                },
                onMonthChange = { month = it },
                modifier = Modifier.padding(top = TeamFisSpacing.sm),
                hasClass = ::hasClass,
            )
            CalendarBar(
                expanded = expanded,
                onToggle = { expanded = !expanded },
                modifier = Modifier.padding(top = TeamFisSpacing.xs),
            )

            val classes = remember(selected) { classesOn(selected) }
            if (classes.isEmpty()) {
                Text(
                    "일정이 없어요",
                    style = TeamFisType.bodySm,
                    color = TeamFisColor.TextMuted,
                    textAlign = TextAlign.Center,
                    modifier = Modifier
                        .fillMaxWidth()
                        .padding(top = TeamFisSpacing.xxxl),
                )
            } else {
                Column(
                    modifier = Modifier.padding(
                        top = TeamFisSpacing.lg,
                        start = TeamFisSpacing.screenHorizontal,
                        end = TeamFisSpacing.screenHorizontal,
                    ),
                    verticalArrangement = Arrangement.spacedBy(TeamFisSpacing.md),
                ) {
                    classes.forEach { item ->
                        ScheduleClassCard(item, onClick = { onClass(item) })
                    }
                }
            }
        }
    }
}

/**
 * 데이터가 붙기 전까지 쓰는 **자리 표시자**다. 서버가 일정을 주면 통째로 걷어낸다.
 * 오늘·내일·사흘 뒤에만 수업이 있는 것으로 둔다 — 점이 찍히는 날과 목록이 어긋나면 안 된다.
 */
private fun hasClass(date: java.time.LocalDate): Boolean {
    val today = java.time.LocalDate.now()
    return date == today || date == today.plusDays(1) || date == today.plusDays(3)
}

private fun classesOn(date: java.time.LocalDate): List<ScheduleClass> =
    if (!hasClass(date)) emptyList() else listOf(
        ScheduleClass("000", date.atTime(10, 0), 60, "얼리버드 20회", "12/20회차", SessionStatus.Done),
        ScheduleClass("000", date.atTime(14, 0), 60, "PT 30회", "12/30회차", SessionStatus.Scheduled),
        ScheduleClass("000", date.atTime(16, 0), 60, "PT 20회", "3/20회차", SessionStatus.Scheduled),
        ScheduleClass("000", date.atTime(18, 30), 60, "얼리버드 10회", "8/10회차", SessionStatus.Scheduled),
    )

/**
 * 캘린더 아래 한 줄 — `펼쳐보기`.
 *
 * 펼치기는 **화살표가 뒤집히며** 캘린더가 그 달로 늘어난다.
 */
@Composable
private fun CalendarBar(expanded: Boolean, onToggle: () -> Unit, modifier: Modifier = Modifier) {
    val interaction = remember { MutableInteractionSource() }
    val arrow by animateFloatAsState(
        targetValue = if (expanded) 180f else 0f,
        animationSpec = TeamFisMotion.base(),
        label = "arrow",
    )

    Row(
        modifier = modifier
            .fillMaxWidth()
            .padding(horizontal = TeamFisSpacing.screenHorizontal - TeamFisSpacing.sm),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        Row(
            modifier = Modifier
                .height(TeamFisSize.chip)
                .clip(TeamFisRadius.full)
                .clickable(interactionSource = interaction, indication = null, onClick = onToggle)
                .padding(horizontal = TeamFisSpacing.sm),
            verticalAlignment = Alignment.CenterVertically,
            horizontalArrangement = Arrangement.spacedBy(2.dp),
        ) {
            Text(
                if (expanded) "접기" else "펼쳐보기",
                style = TeamFisType.bodySm,
                color = TeamFisColor.TextSecondary,
            )
            Icon(
                painter = painterResource(R.drawable.ic_chevron_down),
                contentDescription = null, // 옆 글자가 이름 역할을 한다
                tint = TeamFisColor.TextSecondary,
                modifier = Modifier
                    .size(18.dp)
                    .graphicsLayer { rotationZ = arrow },
            )
        }
    }
}
